src/node/channels.h

```diff
@@ -246,89 +243,293 @@ namespace ccf
       outgoing = false;
     }
 
-    std::vector<uint8_t> get_signed_public()
+    std::vector<uint8_t> sign_key_share(
+      const std::vector<uint8_t>& ks,
+      const std::vector<uint8_t>* extra = nullptr)
     {
-      const auto own_public = ctx.get_own_public();
-      auto signature = network_kp->sign(own_public);
+      std::vector<uint8_t> t = ks;
+
+      if (extra)
+      {
+        t.insert(t.end(), extra->begin(), extra->end());
+      }
+
+      auto signature = node_kp->sign(t);
 
-      // Serialise channel public and network signature and length-prefix both
-      auto space = own_public.size() + signature.size() + 2 * sizeof(size_t);
-      std::vector<uint8_t> serialised_signed_public(space);
-      auto data_ = serialised_signed_public.data();
-      serialized::write(data_, space, own_public.size());
-      serialized::write(data_, space, own_public.data(), own_public.size());
+      // Serialise channel key share, signature, and certificate and
+      // length-prefix them
+      auto space =
+        ks.size() + signature.size() + node_cert.size() + 3 * sizeof(size_t);
+      std::vector<uint8_t> serialised_signed_key_share(space);
+      auto data_ = serialised_signed_key_share.data();
+      serialized::write(data_, space, ks.size());
```

Comment:
For stuff whose length fits in 16bits, we have write_lps.

Reply:
There are no `uint16_t`s here and `write_lps` just assumes that the value fits in 16 bits without checking it. In practice those values will be small, but I'd have to add checks on all the vector sizes myself then.